Lesson2 Task

dmryutov opened the pull request.

dmryutov committed:
Lesson 2 init

dmryutov committed:
Added Google Test

dmryutov committed:
Added Point2D class

dmryutov committed:
Added Box2D and Ray2D classes

dmryutov committed:
Merge pull request #1 from dmryutov/dmryutov

Added Box2D and Ray2D classes

dmryutov committed:
Added tests for Box2D and Ray2D

dmryutov committed:
Added tests for Box2D and Ray2D

dmryutov committed:
All classes updates

- Deleted inheritance from Point2D
- EqualWithEps moved from Point2D to global
- Added assignment operator to Box2D
- Updated Ray2D IntersectBox method
- Updated Ray2D tests

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -0,0 +1,70 @@
+      && m_max.y() > obj.p1().y();
+  }
+private:
+  void PointSwap() {
```
{ выставлена не по codestyle. Перенесите на следующую строку.

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -0,0 +1,70 @@
+  Box2D() = default;
+  Box2D(float x1, float y1, float x2, float y2)
+    : m_min(x1, y1), m_max(x2, y2)
+  { PointSwap(); }
```
Не хватает пустой строки между конструкторами. Рекомендую еще писать так
{
  PointSwap();
}

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -0,0 +1,70 @@
+  { PointSwap(); }
+
+  // Copy constructor
+  Box2D(Box2D const &obj)
```
Здесь и далее между & и именем переменной ставьте пробел

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -0,0 +1,70 @@
+  { PointSwap(); }
+
+  // Getters
+  Point2D &p1() { return m_min; }
```
Имена p1 и p2 непонятны. Лучше использовать min и max, но так как эти имена заняты стандартными функциями, можно назвать, например, boxMin и boxMax.

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -0,0 +1,70 @@
+  // Functionality
+  bool IntersectBox(Box2D const &obj) const
+  {
+    return m_min.x() < obj.p2().x()
```
Этот код выглядит как верный, но его можно улучшить. Проверяйте, не когда выполняется условие, а когда оно не выполняется. Например, если максимум второго бокса меньше минимума первого, то боксы точно не пересекаются.
Вот полезная вам ссылка http://gamemath.com/2011/09/detecting-whether-two-boxes-overlap/

rokuz commented on hometask/lesson2/src/main.cpp:
```diff
@@ -0,0 +1,42 @@
+#include "../src/box2d.hpp"
+#include "../src/ray2d.hpp"
+
+int main() {
```
Скобочку - на новую строку

rokuz commented on hometask/lesson2/src/ray2d.hpp:
```diff
@@ -0,0 +1,84 @@
+  Ray2D(float x1, float y1, float angel)
+    : m_origin(x1, y1), m_angle(angel)
+  {}
+  Ray2D(Point2D p1, float angel)
```
angel - это ангел :) а угол - это angle. Как бы то ни было, предлагаю задавать направление единичным вектором, а не углом поворота.

rokuz commented on hometask/lesson2/src/ray2d.hpp:
```diff
@@ -0,0 +1,84 @@
+  // Functionality
+  bool IntersectBox(Box2D const &obj) const
+  {
+    double tmin = -INFINITY, tmax = INFINITY;
```
Посмотрите класс numeric_limits (#include <climits>) http://en.cppreference.com/w/cpp/types/numeric_limits
Вместо INFINITY правильнее использовать numeric_limits<double>::max() и numeric_limits<double>::lowest()

rokuz commented:
Кроме прочего, забыли скопировать тесты на Point2D.
В целом, очень хорошо! Молодцы!

dmryutov committed:
Codestyle fixes

- Codestyle fixes
- Added Point2D tests
- Updated Box2D IntersectBox method

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -8,53 +8,69 @@ class Box2D
+  Box2D & operator = (Box2D const & obj)
   {
-    if (this == &obj) return *this;
+    if (this == & obj) return *this;
```
А вот здесь как раз все правильно было :) Здесь & это операция взятия адреса

rokuz commented on hometask/lesson2/src/box2d.hpp:
```diff
@@ -0,0 +1,86 @@
+  // Copy constructor
+  Box2D(Box2D const & obj)
+    : m_min(obj.m_min), m_max(obj.m_max)
+  { PointSwap(); }
```
Тогда уж тут тоже надо
{
  PointSwap();
}

dmryutov committed:
Lesson2 Task updates

- EqualWithEps returned to Point2D
- Added assignment operator to Box2D
- Changed Ray2D direction from angle to vector
- Updated Ray2D IntersectBox method
- Updated Ray2D tests
- Codestyle fixes
